### ISSUE TITLE
macro atoms - nebular concentrations fails to converge

### ISSUE THREAD

**Issue by kslong** [sirocco-rt/sirocco#493]
The C345 model is reporting errors associated with nebular concentrations failing to converge. The errors seem to follow bad returns from matrix_ion_populations, and are both failures are likely due to the same cause.   The cause in the nebular concentrations case appears to be associated with w being very small, see below.  The electron and radiation temperatures in this cell are of order 25,000 K so this is not a low temperature problem.  The example below is at the end of the first ionization cycle

![screenshot_970](https://user-images.githubusercontent.com/2530901/55333152-4ef64e80-545c-11e9-9610-6b6fd7a76138.png)

Here is the code around the error message

![screenshot_968](https://user-images.githubusercontent.com/2530901/55332869-d1cad980-545b-11e9-86d9-e2b9f723cdb7.png)
There is something very odd about the error message line, which indicates that t_r is alos small, because I don't see xplasma->t_e in the printout line.

**Comment by kslong** [comment]
All of the other examples of this appear to be similar, in the w is very small.

**Comment by jhmatthews** [comment]
I've made some progress on this one. Pretty sure it's due to the cross-sections error that I described here: #425 

If I run ixvel_c345 for one cycle with 1e5 photons, the error summary is:
```
Error summary: End of program, Thread 0 only
Recurrences --  Description
        7 -- getatomic_data: line input incomplete: %s
       32 -- get_atomicdata: No inner electron yield data for inner cross section %i
        5 -- Ignored %d simple lines for macro-ion %d
        1 -- Ignored %d collision strengths with no matching line transition
        1 -- get_atomicdata: macro atom but no topbase xsection! ion %i z %i istate %i, yet marked as topbase xsection!
        1 -- Macro info for ion doesn't match with xsection!! Could be serious.
        1 -- get_wind_params: zdom[ndom].rmax 0 for wind type %d
        1 -- wind2d: Cell %3d (%2d,%2d) in domain %d has %d corners in wind, but zero volume
        1 -- check_grid: velocity changes by >1,000 km/s in %i cells
        1 -- check_grid: some cells have large changes. Consider modifying zlog_scale or grid dims
      174 -- Solve_matrix: gsl_linalg_LU_solve failure (%d %.3e) for cell %i 
      174 -- matrix_ion_populations: bad return from solve_matrix
      174 -- matrix_ion_populations: Unsolvable matrix! Determinant is zero. Defaulting to no change.
      174 -- ionization_on_the_spot: nebular_concentrations failed to converge
      174 -- ionization_on_the_spot: j %8.2e t_e %8.2e t_r %8.2e w %8.2e nphot %i
        5 -- error_count: This error will no longer be logged: %s
       11 -- The net flow out of simple ion pool (%8.4e) > than the net flow in (%8.4e) in cell %d
Run py_error.py for full error report.
Thread 0 Finalized. All done
xxx Tue Apr  2 14:15:03 2019     38.4 COMPLETE             ixvek_c345
Completed entire program.  The elapsed TIME was 38.449817
```

If I now track the highest ion in a given element and treat that as the continuum in the ionization calculation, the error summary is much better. This ensures I don't try and calculate a photoionization rate for the highest ion. This fix also includes an if statement that leaves out vfky cross-sections for macro-atoms. Error summary:
```
Error summary: End of program, Thread 0 only
Recurrences --  Description
        7 -- getatomic_data: line input incomplete: %s
       31 -- get_atomicdata: No inner electron yield data for inner cross section %i
        5 -- Ignored %d simple lines for macro-ion %d
        1 -- Ignored %d collision strengths with no matching line transition
        1 -- get_wind_params: zdom[ndom].rmax 0 for wind type %d
        1 -- wind2d: Cell %3d (%2d,%2d) in domain %d has %d corners in wind, but zero volume
        1 -- check_grid: velocity changes by >1,000 km/s in %i cells
        1 -- check_grid: some cells have large changes. Consider modifying zlog_scale or grid dims
       11 -- The net flow out of simple ion pool (%8.4e) > than the net flow in (%8.4e) in cell %d
Run py_error.py for full error report.
Thread 0 Finalized. All done
xxx Tue Apr  2 14:10:59 2019     39.8 COMPLETE             ixvek_c345
Completed entire program.  The elapsed TIME was 39.780262
```
This change is in the max_ion_stage branch and can be checked out and used for testing. However I don't think it should be merged to dev without thorough testing as it could mess up the ionization if I've made a mistake and someone should check it carefully.

**Comment by kslong** [comment]
@jhmatthews  I don't see the max_ion_stage branch.  Did you push it back to github?

**Comment by jhmatthews** [comment]
Sorry Knox, pushed now.

**Comment by jhmatthews** [comment]
I'm mystified by this, but for some reason, in this new max_ion_stage branch only, I get good results when I run in serial but terrible errors when I run in parallel - lots of nan macro_pops errors. So the max_ion_stage may not be as good as I'd hoped.

**Comment by jhmatthews** [comment]
Ok, I'd made a silly mistake in my new fix. I've changed that and pushed, so it should now be giving sensible results. Knox, if you like, you can try using this branch with the ixvel_c345 model to see which of the errors it eliminates. I am doing some runs in the background while I work on other things.

**Comment by jhmatthews** [comment]
I think this particular issue has been fixed by #502 - max_ion_stage merge, so I'm closing.
